tools/nl/loadtest/loadtest.py

```diff
@@ -72,9 +95,79 @@ def load_test(query_file, output_file):
   print('')
 
 
+# Asynchronous function to run a query and add result of the run to results
+async def run_query(results, key, session, query):
+  try:
+    print(f'Query {key}: {query}')
+    response = await session.get(_URL + query)
+    # Check the response status code
+    if response.status == 200:
+      # parse the json response and add to the result
+      data = await response.json()
+      print(f'Query {key} completed successfully')
+      debug_timing = data.get('debug', {}).get('timing', {})
+      charts = data.get('charts', [])
+      chart_types = list(set([c['type'] for c in charts]))
+      results[key] = {
+          _RESULT_KEY_STATUS: 'EMPTY' if len(charts) == 0 else 'SUCCESS',
+          _RESULT_KEY_QUERY: query,
+          _RESULT_KEY_NUM_CHARTS: str(len(data['charts'])),
+          _RESULT_KEY_CHART_TYPES: chart_types,
+          _RESULT_KEY_NL_TIME: debug_timing.get('getNlResult', 0),
+          _RESULT_KEY_TILE_TIME: debug_timing.get('getTileResults', 0),
+          _RESULT_KEY_TOTAL_TIME: debug_timing.get('total', 0)
+      }
+    else:
+      print(f'Query {key} failed with status code:', response.status_code)
+      results[key] = {
+          _RESULT_KEY_STATUS: 'FAILURE',
+          _RESULT_KEY_QUERY: query,
+      }
+  except Exception as e:
+    print(f'Error running query {key}: {str(e)}')
+    results[key] = {
+        _RESULT_KEY_STATUS: 'FAILURE',
+        _RESULT_KEY_QUERY: query,
+    }
+
+
+# Runs asynchronous load test where all the queries are requested in parallel
+async def run_async_load_test(query_file, output_file):
+  results = {}
+  result_keys = []
+  async with aiohttp.ClientSession() as session:
```

Comment:
Also, I wonder about using concurrent.futures() instead?

So the same code can be used for single-threaded (if num-workers = 1) or parallel load?

Reply:
done